pkg/cluster/sync.go

```diff
@@ -107,73 +78,159 @@ func (c *Cluster) Sync(newSpec *spec.Postgresql) (err error) {
+
+		desiredSvc := c.generateService(role, &c.Spec)
+		match, reason := k8sutil.SameService(svc, desiredSvc)
+		if match {
+			c.Services[role] = svc
+			return nil
+		}
+		c.logServiceChanges(role, svc, desiredSvc, false, reason)
+
+		if err := c.updateService(role, desiredSvc); err != nil {
+			return fmt.Errorf("could not update %s service to match desired state: %v", role, err)
+		}
+		c.logger.Infof("%s service %q is in the desired state now", role, util.NameFromMeta(desiredSvc.ObjectMeta))
 
 		return nil
+	} else if !k8sutil.ResourceNotFound(err) {
+		return fmt.Errorf("could not get %s service: %v", role, err)
 	}
+	c.Services[role] = nil
 
-	desiredSvc := c.generateService(role, &cSpec)
-	match, reason := c.sameServiceWith(role, desiredSvc)
-	if match {
+	// Service does not exist
+	if role == Replica && !c.Spec.ReplicaLoadBalancer {
 		return nil
 	}
-	c.logServiceChanges(role, c.Services[role], desiredSvc, false, reason)
 
-	if err := c.updateService(role, desiredSvc); err != nil {
-		return fmt.Errorf("could not update %s service to match desired state: %v", role, err)
+	c.logger.Infof("could not find the cluster's %s service", role)
+
+	if svc, err := c.createService(role); err != nil {
+		if k8sutil.ResourceAlreadyExists(err) {
+			c.logger.Infof("%s service %q already exists", role, util.NameFromMeta(svc.ObjectMeta))
+		} else {
+			return fmt.Errorf("could not create missing %s service: %v", role, err)
+		}
+	} else {
+		c.logger.Infof("created missing %s service %q", role, util.NameFromMeta(svc.ObjectMeta))
+		c.Services[role] = svc
 	}
-	c.logger.Infof("%s service %q is in the desired state now", role, util.NameFromMeta(desiredSvc.ObjectMeta))
 
 	return nil
 }
 
-func (c *Cluster) syncEndpoint() error {
-	if c.Endpoint == nil {
-		c.logger.Infof("could not find the cluster's endpoint")
-		ep, err := c.createEndpoint()
-		if err != nil {
-			return fmt.Errorf("could not create missing endpoint: %v", err)
+func (c *Cluster) syncEndpoint(role PostgresRole) error {
+	c.setProcessName("syncing %s endpoint", role)
+
+	ep, err := c.KubeClient.Endpoints(c.Namespace).Get(c.endpointName(role), metav1.GetOptions{})
+	if err == nil {
+		if role == Replica && !c.Spec.ReplicaLoadBalancer {
+			if err := c.deleteEndpoint(role); err != nil {
+				return fmt.Errorf("could not delete %s endpoint", role)
+			}
 		}
-		c.logger.Infof("created missing endpoint %q", util.NameFromMeta(ep.ObjectMeta))
+
+		c.Endpoints[role] = ep
 		return nil
+	} else if !k8sutil.ResourceNotFound(err) {
+		return fmt.Errorf("could not get %s endpoint: %v", role, err)
+	}
+	c.Endpoints[role] = nil
+
+	// Endpoint does not exist
+	if role == Replica && !c.Spec.ReplicaLoadBalancer {
+		return nil
+	}
+
+	c.logger.Infof("could not find the cluster's %s endpoint", role)
+
+	if ep, err := c.createEndpoint(role); err != nil {
+		if k8sutil.ResourceAlreadyExists(err) {
+			c.logger.Infof("%s endpoint %q already exists", role, util.NameFromMeta(ep.ObjectMeta))
```

Comment:
same issue as with the services, c.Endpoints will have no knowledge of the existing endpoint.